File: services/calc-engine/routers/analysis.py
"""
Analysis router — handles POST /analysis requests from the Fastify API.
Route handlers call services and calculations. No business logic here.
"""

import logging
from fastapi import APIRouter
from pydantic import BaseModel, Field

from models.schemas import (
    PropertyInput,
    FinancingInput,
    RentalEstimate,
    AnalysisOutput,
    InvestmentMetricsOutput,
    DealScoreOutput,
    DealScoreBreakdownOutput,
    ComponentMaxes,
    SunScoutOutput,
)
from sunscout.sun_path import calculate_sun_hours
from constants.rates import get_maintenance_rate
from calculations.mortgage import calculate_monthly_payment
from calculations.closing_costs import estimate_closing_costs
from calculations.investment import (
    calculate_noi,
    calculate_cap_rate,
    calculate_cash_flow_monthly,
    calculate_dscr,
    calculate_grm,
    calculate_cash_on_cash,
    calculate_break_even_rent,
)
from calculations.deal_score import calculate_deal_score, to_display_score
from calculations.sanity import sanity_check_metrics
from extraction.regex_rules import extract_regex_flags
from extraction.haiku_extraction import extract_flags_with_haiku
from extraction.logic_gate import merge_flags, MergedFlag

logger = logging.getLogger(__name__)

router = APIRouter()

# ── CMHC defaults (nightly scraper not yet wired — Week 5–6) ─────────────────
_DEFAULT_CMHC_VACANCY_RATE: float = 0.02
_DEFAULT_RENTAL_DOM: int = 21
_DEFAULT_RENT_TREND: str = "flat"

# Points deducted from the deal score per confirmed STANDARD red flag
_DEDUCTION_PER_RED_FLAG: int = 5

# Severe dealbreakers — these GATE the score's ceiling instead of deducting
# points (spec §10a). Each must have a deterministic regex floor (verified in
# extraction/regex_rules_test.py) so the gate never runs on nothing when Haiku
# is unavailable. Modes that show the investment score (investor/landlord) apply
# the gate; personal/tenant don't use this score.
_SEVERE_FLAGS: frozenset[str] = frozenset(
    {
        "grow_op_history",
        "flooding_history",
        "illegal_unit_risk",
        "special_assessment_risk",
    }
)
_GATED_MODES: frozenset[str] = frozenset({"investor", "landlord"})


class AnalysisRequest(BaseModel):
    """Combined request body for the analysis endpoint."""

    property_data: PropertyInput
    financing: FinancingInput
    rental: RentalEstimate
    description: str | None = (
        None  # raw listing description; run through extraction pipeline
    )
    # Real per-city CMHC vacancy rate (decimal, e.g. 0.018). Supplied by the
    # Fastify API from cmhcService; falls back to the default when omitted so
    # the demand score reflects the actual market, not a flat assumption.
    cmhc_vacancy_rate: float | None = None
    # Flag IDs the user has dismissed for this analysis (from flag_overrides).
    # Dismissed red flags are still returned in risk_flags (so the UI can show
    # them greyed out) but no longer deduct from the deal score on re-run.
    dismissed_flag_ids: list[str] = Field(default_factory=list)
    # Report mode — only investor/landlord apply the investment severe-gate
    # (personal uses HomeScore, tenant has no deal score). Spec §10a.
    mode: str = "investor"


@router.post("/", response_model=AnalysisOutput)
async def run_analysis(body: AnalysisRequest) -> AnalysisOutput:
    """
    Run a full investment analysis for a property.

    Steps:
      1. Derive maintenance rate from build year.
      2. Compute mortgage payment and closing costs.
      3. Compute all investment metrics (NOI, cap rate, cash flow, DSCR, GRM, CoC,
         break-even rent).
      4. Compute deal score.
      5. Run sanity checks; log any failures.
      6. Return AnalysisOutput to the Fastify API.

    Args:
        body: Combined request with property_data, financing, and rental inputs.

    Returns:
        Full AnalysisOutput with metrics, deal_score, risk_flags, and warning flag.
    """
    prop = body.property_data
    fin = body.financing
    rental = body.rental

    # ── 1. Maintenance rate ───────────────────────────────────────────────────
    maintenance_rate = get_maintenance_rate(prop.year_built)

    # ── 2. Mortgage + closing costs ───────────────────────────────────────────
    down_payment = float(prop.price) * fin.down_payment_pct
    mortgage_amount = float(prop.price) - down_payment

    mortgage_payment_monthly = calculate_monthly_payment(
        principal=mortgage_amount,
        annual_rate=fin.mortgage_rate,
        amortization_years=fin.amortization_years,
    )
    annual_debt_service = mortgage_payment_monthly * 12

    closing = estimate_closing_costs(
        purchase_price=float(prop.price),
        is_toronto=prop.is_toronto,
    )

    # ── 3. Investment metrics ─────────────────────────────────────────────────
    monthly_rent = float(rental.mid)
    gross_annual_rent = monthly_rent * 12
    condo_fee_monthly = float(prop.condo_fee_monthly or 0)

    noi = calculate_noi(
        gross_annual_rent=gross_annual_rent,
        annual_taxes=float(prop.annual_taxes),
        insurance_value=float(prop.price),
        condo_fee_annual=condo_fee_monthly * 12,
        maintenance_rate=maintenance_rate,
        property_value=float(prop.price),
        include_management=fin.include_management_fee,
    )

    cap_rate = calculate_cap_rate(noi=noi, purchase_price=float(prop.price))

    cash_flow_monthly = calculate_cash_flow_monthly(
        monthly_rent=monthly_rent,
        mortgage_payment=mortgage_payment_monthly,
        annual_taxes=float(prop.annual_taxes),
        insurance_value=float(prop.price),
        condo_fee_monthly=condo_fee_monthly,
        maintenance_rate=maintenance_rate,
        property_value=float(prop.price),
        include_management=fin.include_management_fee,
    )
    cash_flow_annual = cash_flow_monthly * 12

    dscr = calculate_dscr(noi=noi, annual_debt_service=annual_debt_service)

    grm = (
        calculate_grm(
            purchase_price=float(prop.price), annual_gross_rent=gross_annual_rent
        )
        if gross_annual_rent > 0
        else 0.0
    )

    total_cash_invested = down_payment + closing["total"]
    cash_on_cash = calculate_cash_on_cash(
        annual_cash_flow=cash_flow_annual,
        total_cash_invested=total_cash_invested,
    )

    break_even_rent = calculate_break_even_rent(
        mortgage_payment=mortgage_payment_monthly,
        annual_taxes=float(prop.annual_taxes),
        insurance_value=float(prop.price),
        condo_fee_monthly=condo_fee_monthly,
        maintenance_rate=maintenance_rate,
        property_value=float(prop.price),
        include_management=fin.include_management_fee,
    )

    # ── 3b. Listing description extraction pipeline ───────────────────────────
    # Runs regex first (deterministic), then Haiku for gray areas.
    # Logic gate merges and applies confidence thresholds.
    # Only validated red flags (85%+ confidence) deduct from the deal score.
    merged_flags: list = []
    risk_flag_deductions: float = 0.0
    severe_flag_count: int = 0

    dismissed_flags = set(body.dismissed_flag_ids)

    if body.description:
        try:
            regex_flags = extract_regex_flags(body.description)
            haiku_flags = await extract_flags_with_haiku(body.description)
            merged_flags = merge_flags(regex_flags, haiku_flags)

            # Active = red and not dismissed. Severe flags GATE the ceiling;
            # standard reds deduct −5 each (capped in calculate_deal_score).
            active_reds = [
                f
                for f in merged_flags
                if f.severity == "red" and f.flag_id not in dismissed_flags
            ]
            severe_flag_count = sum(
                1 for f in active_reds if f.flag_id in _SEVERE_FLAGS
            )
            standard_red_count = len(active_reds) - severe_flag_count
            risk_flag_deductions = float(standard_red_count * _DEDUCTION_PER_RED_FLAG)
        except Exception as exc:  # noqa: BLE001 — non-fatal; rest of report still loads
            logger.error("Extraction pipeline failed for %s: %s", prop.address, exc)
            merged_flags = []
            risk_flag_deductions = 0.0
            severe_flag_count = 0

    # ── 3c. Structural flag — condo with unknown fee ──────────────────────────
    # The calculation uses $0 when condo_fee_monthly is not provided, which may
    # silently understate carrying costs by hundreds per month for condos.
    # Surface an amber warning so the user can correct it in the financing sliders.
    if prop.property_type == "condo" and not prop.condo_fee_known:
        condo_fee_flag = MergedFlag(
            flag_id="condo_fee_unknown",
            severity="amber",
            confidence=100,
            evidence="Condo fee not provided — costs calculated assuming $0/month",
            source="structural",
        )
        if not any(f.flag_id == "condo_fee_unknown" for f in merged_flags):
            merged_flags.append(condo_fee_flag)

    # ── 4. Deal score ─────────────────────────────────────────────────────────
    # Use the real per-city vacancy rate when the API supplies it; the DOM and
    # rent-trend inputs stay on defaults until a data source exists for them.
    vacancy_rate = (
        body.cmhc_vacancy_rate
        if body.cmhc_vacancy_rate is not None
        else _DEFAULT_CMHC_VACANCY_RATE
    )
    # The severe gate only applies to modes that show this investment score.
    gate_count = severe_flag_count if body.mode in _GATED_MODES else 0
    score_result = calculate_deal_score(
        cap_rate=cap_rate,
        cash_flow_monthly=cash_flow_monthly,
        cash_on_cash=cash_on_cash,
        dscr=dscr,
        cmhc_vacancy_rate=vacancy_rate,
        rental_days_on_market=_DEFAULT_RENTAL_DOM,
        rent_trend=_DEFAULT_RENT_TREND,
        risk_flag_deductions=risk_flag_deductions,
        severe_flag_count=gate_count,
    )

    br = score_result["breakdown"]
    cm = br["component_maxes"]
    deal_score = DealScoreOutput(
        total=score_result["total"],
        display_total=to_display_score(int(score_result["total"])),
        verdict=score_result["verdict"],
        breakdown=DealScoreBreakdownOutput(
            cap_rate=br["cap_rate"],
            cash_flow=br["cash_flow"],
            cash_on_cash=br["cash_on_cash"],
            dscr=br["dscr"],
            demand=br["demand"],
            subtotal=br["subtotal"],
            deduction=br["deduction"],
            component_maxes=ComponentMaxes(
                cap_rate=cm["cap_rate"],
                cash_flow=cm["cash_flow"],
                cash_on_cash=cm["cash_on_cash"],
                dscr=cm["dscr"],
                demand=cm["demand"],
            ),
        ),
    )

    # ── 5. Sanity checks ──────────────────────────────────────────────────────
    sanity_warnings = sanity_check_metrics(
        cap_rate=cap_rate,
        monthly_rent=monthly_rent,
        purchase_price=float(prop.price),
        dscr=dscr,
        break_even_rent=break_even_rent,
        deal_score=deal_score.total,
        cash_flow_monthly=cash_flow_monthly,
    )
    has_sanity_warnings = len(sanity_warnings) > 0

    if sanity_warnings:
        for warning in sanity_warnings:
            logger.warning("Sanity check failed for %s: %s", prop.address, warning)

    # ── 6. SunScout (non-fatal — skipped when lat/lng unavailable) ───────────
    sun_scout_result: SunScoutOutput | None = None
    if prop.lat is not None and prop.lng is not None:
        try:
            ss = calculate_sun_hours(prop.lat, prop.lng)
            bedroom_main = ss.window_hours.get("bedroom_main", {})
            monthly_hours = [bedroom_main.get(m, 0.0) for m in range(1, 13)]
            sun_scout_result = SunScoutOutput(
                annual_peak_sun_hours=ss.annual_peak_sun_hours,
                summer_daily_hours=ss.summer_daily_hours,
                winter_daily_hours=ss.winter_daily_hours,
                seasonal_grid=ss.seasonal_grid,
                monthly_hours=monthly_hours,
                sun_score=ss.sun_score,
                verdict=ss.verdict,
            )
        except Exception as exc:  # noqa: BLE001
            logger.error("SunScout failed for %s: %s", prop.address, exc)

    # ── 7. Assemble and return output ─────────────────────────────────────────
    metrics = InvestmentMetricsOutput(
        cash_flow_monthly=round(cash_flow_monthly, 2),
        cash_flow_annual=round(cash_flow_annual, 2),
        cap_rate=round(cap_rate, 6),
        cash_on_cash_return=round(cash_on_cash, 6),
        dscr=round(dscr, 4),
        grm=round(grm, 2),
        noi=round(noi, 2),
        mortgage_payment_monthly=round(mortgage_payment_monthly, 2),
        down_payment=round(down_payment, 2),
        mortgage_amount=round(mortgage_amount, 2),
        amortization_years=fin.amortization_years,
        mortgage_rate=fin.mortgage_rate,
        break_even_rent=round(break_even_rent, 2),
        closing_costs_total=round(closing["total"], 2),
        ltt_provincial=round(closing["ltt_provincial"], 2),
        ltt_municipal=round(closing["ltt_municipal"], 2),
        has_sanity_warnings=has_sanity_warnings,
    )

    # Serialise merged flags into the response — only flags above amber threshold
    serialised_flags: list[dict[str, object]] = [
        {
            "flag_id": f.flag_id,
            "severity": f.severity,
            "confidence": f.confidence,
            "evidence": f.evidence,
            "source": f.source,
        }
        for f in merged_flags
    ]

    return AnalysisOutput(
        metrics=metrics,
        deal_score=deal_score,
        risk_flags=serialised_flags,
        has_sanity_warnings=has_sanity_warnings,
        sun_scout=sun_scout_result,
    )

File: apps/api/src/routes/analysis.ts
/**
 * Analysis route — accepts { token, mode } from the frontend, runs the
 * full 9-step pipeline (flags → calc engine → geocode → walk score →
 * narrative → assemble), and returns { token, analysis }.
 *
 * Registered in app.ts with prefix "/analysis":
 *   await fastify.register(import('./routes/analysis'), { prefix: '/analysis' })
 * Fastify dynamic import requires a default export.
 */

import { type FastifyInstance } from 'fastify'
import { makeError } from '../types/api'
import type {
  InvestmentMetrics,
  DealScore,
  DealScoreBreakdown,
  Analysis,
  ReportMode,
  WalkScoreResult,
} from '../types/analysis'
import {
  getListingByToken,
  updateAnalysisStatus,
  updateAnalysisByToken,
  fetchRentalComps,
  getFlagOverrides,
  getNearbySchools,
} from '../services/supabaseService'
import { generateNarrative, type NarrativeInput } from '../services/anthropicService'
import { geocodeAddress } from '../services/mapboxService'
import { getWalkScore } from '../services/walkScoreService'
import { getNearbyDistances } from '../services/googlePlacesService'
import { getNeighbourhoodStats } from '../services/statsCanService'
import { getVacancyRateByCity } from '../services/cmhcService'
import { getMortgageRate } from '../services/bankOfCanadaService'
import { flagLabel } from '../constants/flagLabels'
import { estimateAnnualTaxes } from '../constants/propertyTaxRates'
import { RENT_BOUNDS, CALC_ENGINE_TIMEOUT_MS } from '../constants/thresholds'
import { serializeError, isTimeoutError } from '../lib/http'
import {
  RENT_TO_PRICE_MONTHLY,
  FALLBACK_RENT_BAND,
  DEFAULT_RENT_MONTHLY,
} from '../constants/valuation'
import { estimateValueFromRent } from '../constants/marketCapRates'

const CALC_ENGINE_URL = process.env.CALC_ENGINE_URL ?? 'http://localhost:8000'

// Mortgage rate is overridden per-request by the live Bank of Canada rate
// (via bankOfCanadaService → calc engine /rates/mortgage). 0.0479 is the
// hardcoded fallback used by the calc engine when the live fetch fails — kept
// here too so tests don't need to mock the rate service.
const FINANCING_DEFAULTS = {
  down_payment_pct: 0.2,
  mortgage_rate: 0.0479,
  amortization_years: 25,
  include_management_fee: false,
}

const VALID_MODES: readonly string[] = ['investor', 'personal', 'tenant', 'landlord']

// ── Python response types (snake_case from calc engine) ───────────────────────

interface PyComponentMaxes {
  cap_rate: number
  cash_flow: number
  cash_on_cash: number
  dscr: number
  demand: number
}

interface PyDealScoreBreakdown {
  cap_rate: number
  cash_flow: number
  cash_on_cash: number
  dscr: number
  demand: number
  subtotal: number
  deduction: number
  component_maxes: PyComponentMaxes
}

interface PyDealScore {
  total: number
  display_total: number
  verdict: string
  breakdown: PyDealScoreBreakdown
}

interface PyInvestmentMetrics {
  cash_flow_monthly: number
  cash_flow_annual: number
  cap_rate: number
  cash_on_cash_return: number
  dscr: number
  grm: number
  noi: number
  mortgage_payment_monthly: number
  down_payment: number
  mortgage_amount: number
  amortization_years: number
  mortgage_rate: number
  break_even_rent: number
  closing_costs_total: number
  ltt_provincial: number
  ltt_municipal: number
  has_sanity_warnings: boolean
}

interface PyRiskFlag {
  /** Calc engine uses snake_case `flag_id`; older callers may send `id`. */
  flag_id?: string
  id?: string
  severity?: string
  /** Per-mode severity tier from the flag matrix: severe|red|amber. */
  tier?: string
  /** Calc engine doesn't send a label — orchestrator resolves it via flagLabel(). */
  label?: string
  evidence?: string | null
  confidence?: number
  [key: string]: unknown
}

export interface PySunScout {
  annual_peak_sun_hours: number
  summer_daily_hours: number
  winter_daily_hours: number
  seasonal_grid: { Dec: number; Mar: number; Jun: number; Sep: number }
  monthly_hours: number[]
  sun_score: number
  verdict: string
}

interface PyAnalysisOutput {
  metrics: PyInvestmentMetrics
  deal_score: PyDealScore
  risk_flags: PyRiskFlag[]
  has_sanity_warnings: boolean
  /** Present only when lat/lng were sent and the sun-path calc succeeded. */
  sun_scout?: PySunScout | null
}

// ── snake_case → camelCase output transforms ──────────────────────────────────

export function toSunScout(py: PySunScout | null | undefined): Analysis['sunScout'] {
  if (py == null) return null
  return {
    annualPeakSunHours: py.annual_peak_sun_hours,
    summerDailyHours: py.summer_daily_hours,
    winterDailyHours: py.winter_daily_hours,
    seasonalGrid: py.seasonal_grid,
    monthlyHours: py.monthly_hours,
    sunScore: py.sun_score,
    verdict: py.verdict as NonNullable<Analysis['sunScout']>['verdict'],
  }
}

function toMetrics(py: PyInvestmentMetrics): InvestmentMetrics {
  return {
    cashFlowMonthly: py.cash_flow_monthly,
    cashFlowAnnual: py.cash_flow_annual,
    capRate: py.cap_rate,
    cashOnCashReturn: py.cash_on_cash_return,
    dscr: py.dscr,
    grm: py.grm,
    noi: py.noi,
    mortgagePaymentMonthly: py.mortgage_payment_monthly,
    downPayment: py.down_payment,
    mortgageAmount: py.mortgage_amount,
    amortizationYears: py.amortization_years,
    mortgageRate: py.mortgage_rate,
    breakEvenRent: py.break_even_rent,
    closingCostsTotal: py.closing_costs_total,
    lttProvincial: py.ltt_provincial,
    lttMunicipal: py.ltt_municipal,
    hasSanityWarnings: py.has_sanity_warnings,
  }
}

function toDealScore(py: PyDealScore): DealScore {
  const br = py.breakdown
  const cm = br.component_maxes
  const breakdown: DealScoreBreakdown = {
    capRate: br.cap_rate,
    cashFlow: br.cash_flow,
    cashOnCash: br.cash_on_cash,
    dscr: br.dscr,
    demand: br.demand,
    subtotal: br.subtotal,
    deduction: br.deduction,
    componentMaxes: {
      capRate: cm.cap_rate,
      cashFlow: cm.cash_flow,
      cashOnCash: cm.cash_on_cash,
      dscr: cm.dscr,
      demand: cm.demand,
    },
  }
  return {
    total: py.total,
    displayTotal: py.display_total,
    verdict: py.verdict as DealScore['verdict'],
    breakdown,
  }
}

// ── Route ─────────────────────────────────────────────────────────────────────

async function analysisRoutes(fastify: FastifyInstance): Promise<void> {
  fastify.post<{ Body: { token?: string; mode?: string } }>('/', async (req, reply) => {
    const { token, mode: modeRaw } = req.body ?? {}

    // Step 1 — validate input
    if (!token) {
      return reply
        .code(400)
        .send(makeError('MISSING_TOKEN', 'A token is required to run analysis.'))
    }

    if (!modeRaw || !VALID_MODES.includes(modeRaw)) {
      return reply
        .code(400)
        .send(
          makeError('INVALID_MODE', 'Mode must be one of: investor, personal, tenant, landlord.')
        )
    }

    const mode = modeRaw as ReportMode

    try {
      // Step 2 — look up listing
      const listing = await getListingByToken(token)
      if (!listing) {
        return reply.code(404).send(makeError('NOT_FOUND', 'Analysis not found or has expired.'))
      }

      // Step 3 — mark processing
      await updateAnalysisStatus(token, 'processing')

      // (The listing description is forwarded to the calc engine in the
      // payload below — the calc engine runs the full extraction pipeline
      // (regex + Haiku merge + confidence gating) and applies red-flag
      // deductions to the deal score. Doing extraction here too would be
      // duplicate work and the result would be ignored.)

      // Step 4 — fetch rental comps from nightly-scraped rental_listings.
      // Falls back to a low-confidence estimate from the listing's own rent
      // (or the price-based proxy) when the FSA has no comps yet.
      const comps = await fetchRentalComps(listing.postalCode, listing.beds).catch(() => null)

      const rentalFallback =
        listing.rentMonthly ?? Math.round((listing.price ?? 0) * RENT_TO_PRICE_MONTHLY)
      const rentalForCalc = comps ?? {
        low: Math.round(rentalFallback * (1 - FALLBACK_RENT_BAND)),
        mid: rentalFallback,
        high: Math.round(rentalFallback * (1 + FALLBACK_RENT_BAND)),
        compCount: 0,
        confidence: 'low' as const,
      }

      // Rent plausibility gate (decision 2026-07-01): a mid outside
      // $500–$10,000/mo means the listing carried no usable rent/price (or a
      // legacy row stored a unit error) — scoring it would produce confident
      // garbage. Fail the analysis with a friendly message instead.
      //
      // The gate only applies to OBSERVED rents (the listing's own rent or a
      // comps mid). A for-sale listing with no comps derives its mid from the
      // asking price (~6% gross-yield proxy) — a $3.5M home legitimately
      // proxies to >$10k/mo and must not hard-fail (live bug 2026-07-02);
      // the calc engine's sanity bounds still flag extreme proxy rents. A
      // rent-less FOR-RENT listing has no price either, so its $0 mid still
      // fails the gate as designed.
      const midIsSalePriceProxy =
        comps == null && listing.rentMonthly == null && (listing.price ?? 0) > 0
      if (
        !midIsSalePriceProxy &&
        (rentalForCalc.mid < RENT_BOUNDS.MIN_MONTHLY || rentalForCalc.mid > RENT_BOUNDS.MAX_MONTHLY)
      ) {
        await updateAnalysisStatus(token, 'failed')
        return reply
          .code(422)
          .send(
            makeError(
              'RENT_OUT_OF_BOUNDS',
              'The rent on this listing looks implausible — check the listing or enter details manually.'
            )
          )
      }

      // Step 4b — live mortgage rate (falls back to FINANCING_DEFAULTS.mortgage_rate
      // when the BoC service is unreachable or returns 'fallback').
      const liveRate = await getMortgageRate().catch(() => null)
      const financingForCalc = {
        ...FINANCING_DEFAULTS,
        mortgage_rate: liveRate?.rate ?? FINANCING_DEFAULTS.mortgage_rate,
      }

      // Step 5 — build calc engine payload
      // For for-rent listings (tenant/landlord modes), listing.price is null.
      // Estimate value from rent via the per-city cap rate + residual-expense
      // ratio (NOI / capRate), subtracting actual tax + condo fee when present
      // so they're never double-counted. Keeps the calc engine able to score.
      const estimatedPrice =
        listing.price ??
        estimateValueFromRent({
          rentMonthly: listing.rentMonthly ?? DEFAULT_RENT_MONTHLY,
          city: listing.city,
          propertyType: listing.propertyType,
          annualTaxes: listing.annualTaxes,
          condoFeeMonthly: listing.condoFeeMonthly,
        })

      // Estimate annual taxes from price + city when the scraper couldn't
      // find the actual value. Defaulting to 0 understated carrying costs
      // by $400–800/mo on a typical Ontario property.
      const annualTaxesForCalc =
        listing.annualTaxes ?? estimateAnnualTaxes(estimatedPrice, listing.city)

      // Real per-city CMHC vacancy rate — feeds both the deal score's demand
      // component (calc engine) and the narrative, so they stay consistent.
      const cmhcVacancyRate = getVacancyRateByCity(listing.city)

      // Flags the user previously dismissed for this analysis — forwarded so a
      // re-run drops their score deduction. Non-essential: any failure (or a
      // missing value) defaults to no dismissals; the analysis never fails here.
      let dismissedFlagIds: string[] = []
      try {
        dismissedFlagIds = (await getFlagOverrides(token)) ?? []
      } catch {
        dismissedFlagIds = []
      }

      // Geocode BEFORE the calc engine call — lat/lng in property_data is what
      // makes the calc engine's SunScout (sun-path) branch fire. Non-fatal:
      // null coords just skip SunScout and the real map.
      const coords = await geocodeAddress(listing.address)

      const calcPayload = {
        // Forwarded so the calc engine runs the extraction pipeline and
        // deducts from the deal score for confirmed red flags.
        description: listing.description ?? null,
        property_data: {
          address: listing.address,
          province: listing.province,
          price: estimatedPrice,
          annual_taxes: annualTaxesForCalc,
          condo_fee_monthly: listing.condoFeeMonthly,
          condo_fee_known: listing.condoFeeKnown,
          beds: listing.beds,
          baths: listing.baths,
          sqft: listing.sqft,
          year_built: listing.yearBuilt,
          property_type: listing.propertyType,
          is_toronto: listing.city === 'Toronto',
          lat: coords?.lat ?? null,
          lng: coords?.lng ?? null,
        },
        financing: financingForCalc,
        rental: {
          low: rentalForCalc.low,
          mid: rentalForCalc.mid,
          high: rentalForCalc.high,
          comp_count: rentalForCalc.compCount,
          confidence: rentalForCalc.confidence,
          postal_code: listing.postalCode,
        },
        cmhc_vacancy_rate: cmhcVacancyRate,
        dismissed_flag_ids: dismissedFlagIds,
        mode,
      }

      // Step 6 — call calc engine (explicit timeout: survives the public-edge
      // limit and never hangs undici's default; shares the scrape path's fix).
      let pyResponse: Response
      try {
        pyResponse = await fetch(`${CALC_ENGINE_URL}/analysis/`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(calcPayload),
          signal: AbortSignal.timeout(CALC_ENGINE_TIMEOUT_MS.ANALYSIS),
        })
      } catch (err) {
        fastify.log.error(
          { err: serializeError(err), timedOut: isTimeoutError(err) },
          'Calc engine unreachable'
        )
        await updateAnalysisStatus(token, 'failed')
        return reply
          .code(503)
          .send(
            makeError(
              'CALC_ENGINE_UNAVAILABLE',
              'Analysis service is temporarily unavailable — try again in a moment.'
            )
          )
      }

      if (!pyResponse.ok) {
        const raw = await pyResponse.text().catch(() => '')
        fastify.log.error({ status: pyResponse.status, body: raw }, 'Calc engine returned error')
        await updateAnalysisStatus(token, 'failed')
        return reply
          .code(500)
          .send(
            makeError(
              'CALC_ENGINE_ERROR',
              'Analysis failed due to an internal error — try again shortly.'
            )
          )
      }

      const pyData = (await pyResponse.json()) as PyAnalysisOutput

      // Step 7 — walk score (uses the coordinates geocoded before the calc call)
      const walkScore: WalkScoreResult | null = coords
        ? await getWalkScore(listing.address, coords.lat, coords.lng)
        : null

      // Step 7b — nearest schools from the schools table (spec: 3 per level,
      // distance-ranked, catchment NOT verified). Null until the EQAO/Fraser
      // CSV is loaded or when geocoding failed — the report shows "data
      // pending", and this lights up the moment the table has rows.
      let schools: Analysis['schools'] = null
      if (coords) {
        try {
          schools = (await getNearbySchools(coords.lat, coords.lng)) ?? null
        } catch {
          schools = null
        }
      }

      // Step 7c — nearby amenity distances (Google Places): transit, grocery,
      // highway on-ramp, pharmacy. Null when no coords / key missing / no results.
      let nearbyDistances: Analysis['nearbyDistances'] = null
      if (coords) {
        try {
          const d = await getNearbyDistances(coords.lat, coords.lng)
          nearbyDistances = d.length > 0 ? d : null
        } catch {
          nearbyDistances = null
        }
      }

      // Step 7d — census stats (StatsCan) for the listing's FSA: median household
      // income + 5-year population growth. Null when the FSA isn't in the table.
      let neighbourhoodStats: Analysis['neighbourhoodStats'] = null
      try {
        neighbourhoodStats = await getNeighbourhoodStats(listing.postalCode)
      } catch {
        neighbourhoodStats = null
      }

      // Step 8 — generate narrative (never throws)
      // The Python calc engine returns flag_id (not id) and no label; resolve
      // human-readable labels here for both the narrative + the UI payload.
      const resolvedFlags = pyData.risk_flags.map((f) => {
        const id = String(f.flag_id ?? f.id ?? '')
        return {
          id,
          severity: (f.severity as 'red' | 'amber') ?? 'amber',
          tier: (f.tier as 'severe' | 'red' | 'amber') ?? 'amber',
          label: flagLabel(id),
          evidence: (f.evidence as string | null) ?? null,
          confidence: Number(f.confidence ?? 0),
        }
      })

      const flagLabels = resolvedFlags
        .map((f) => f.label)
        .filter(Boolean)
        .join(', ')

      const narrativeInput: NarrativeInput = {
        mode,
        tier: 'free',
        address: listing.address,
        price: listing.price,
        capRate: pyData.metrics.cap_rate,
        cashFlowMonthly: pyData.metrics.cash_flow_monthly,
        cashFlowAnnual: pyData.metrics.cash_flow_annual,
        cashOnCash: pyData.metrics.cash_on_cash_return,
        dscr: pyData.metrics.dscr,
        dealScore: pyData.deal_score.total,
        dealVerdict: pyData.deal_score.verdict,
        rentMid: rentalForCalc.mid,
        compCount: rentalForCalc.compCount,
        rentConfidence: rentalForCalc.confidence,
        breakEvenRent: pyData.metrics.break_even_rent,
        condoFeeMonthly: listing.condoFeeMonthly,
        condoFeeKnown: listing.condoFeeKnown,
        rentTrend: 'flat',
        vacancyRate: cmhcVacancyRate,
        riskFlagSummary: flagLabels || undefined,
        // Tenant-verdict inputs — without these the tenant prompt saw only $0
        // asking/range and wrongly concluded "no market data available" even
        // when §01 showed real comps.
        askingRent: listing.rentMonthly ?? undefined,
        rentLow: rentalForCalc.low,
        rentHigh: rentalForCalc.high,
      }

      const narrative = await generateNarrative(narrativeInput)

      // Step 9 — assemble Analysis object
      const analysis: Analysis = {
        id: token,
        token,
        mode,
        createdAt: new Date().toISOString(),
        metrics: toMetrics(pyData.metrics),
        dealScore: toDealScore(pyData.deal_score),
        rentalComps: comps
          ? {
              low: comps.low,
              mid: comps.mid,
              high: comps.high,
              compCount: comps.compCount,
              confidence: comps.confidence,
              postalCode: listing.postalCode,
            }
          : null,
        riskFlags: resolvedFlags,
        narrative,
        walkScore,
        neighbourhood: null,
        nearbyDistances,
        neighbourhoodStats,
        sunScout: toSunScout(pyData.sun_scout),
        coordinates: coords != null ? { lat: coords.lat, lng: coords.lng } : null,
        schools,
        hasSanityWarnings: pyData.has_sanity_warnings,
      }

      // Step 10 — save and return
      await updateAnalysisByToken(token, analysis)
      return reply.send({ token, analysis })
    } catch (err) {
      fastify.log.error({ err }, 'Unexpected error in POST /analysis')
      await updateAnalysisStatus(token, 'failed').catch(() => {})
      return reply.code(500).send(makeError('INTERNAL_ERROR', 'Something went wrong — try again.'))
    }
  })
}

export default analysisRoutes
